packages/dd-trace/src/encode/agentless-ci-visibility.js

```diff
@@ -0,0 +1,245 @@
+'use strict'
+const tracerVersion = require('../../lib/version')
+const { truncateSpan, normalizeSpan } = require('./tags-processors')
+const Chunk = require('./chunk')
+const log = require('../log')
+const { AgentEncoder } = require('./0.4')
+
+const ENCODING_VERSION = 1
+
+function formatSpan (span) {
+  return {
+    type: span.type === 'test' ? 'test' : 'span',
+    version: ENCODING_VERSION,
+    content: normalizeSpan(truncateSpan(span))
+  }
+}
+
+class AgentlessCiVisibilityEncoder extends AgentEncoder {
+  constructor ({ runtimeId, service, env }) {
+    super(...arguments)
+    this._events = []
+    this.runtimeId = runtimeId
+    this.service = service
+    this.env = env
+    this._traceBytes = new Chunk()
+    this._stringBytes = new Chunk()
+    this._stringCount = 0
+    this._stringMap = {}
+
+    this.reset()
+  }
+
+  count () {
+    return this._events.length
+  }
+
+  append (trace) {
+    this._events = this._events.concat(trace)
+  }
+
+  _encodeValue (bytes, value) {
```

Comment:
Why is this overridden?

Reply:
as I mention in https://github.com/DataDog/dd-trace-js/pull/1880/files#r831448048, this `_encodeValue` is now being called for values in the high level payload dictionary, so we also need to encode other objects, like dictionaries and arrays. What I do here is add a case `'object'`, which `0.4` does not need because it is only encoding span metadata

Reply:
Commented on that elsewhere, but we can probably avoid calling this entirely in the first place.

Reply:
done this in https://github.com/DataDog/dd-trace-js/pull/1880/commits/4b6286423f62b5d0343bd9ecc81a9f243a8191fc